File: leetcode-04/solutions/python3/0875.py
class Solution:
    def minEatingSpeed(self, piles: List[int], H: int) -> int:
        l = 1
        r = max(piles) + 1

        while l < r:
            m = (l + r) >> 1
            hour = 0
            for pile in piles:
                hour += (pile - 1) // m + 1
            if hour <= H:
                r = m
            else:
                l = m + 1

        return l

File: leetcode-04/solutions/cpp/0875.cpp
class Solution {
 public:
  int minEatingSpeed(vector<int>& piles, int H) {
    int l = 1;
    int r = *max_element(piles.begin(), piles.end()) + 1;

    while (l < r) {
      int m = (l + r) >> 1;
      int hour = 0;
      for (int pile : piles) hour += (pile - 1) / m + 1;
      if (hour <= H)
        r = m;
      else
        l = m + 1;
    }

    return l;
  }
};